rust/kernel/net/ethtool.rs

```diff
@@ -0,0 +1,246 @@
+    };
+    ($($i:ident),+) => {
+        #[allow(clippy::needless_update)]
+        const TO_USE: kernel::net::ethtool::EthToolToUse =
+            $crate::net::ethtool::EthToolToUse {
+                $($i: true),+ ,
+                ..$crate::net::ethtool::ETH_TOOL_USE_NONE
+            };
+    };
+}
+
+/// Operations table needed for ethtool.
+/// [`Self::TO_USE`] defines which functions are implemented by this type.
+pub trait EthToolOps<T: NetDeviceAdapter>: Send + Sync + Sized {
+    /// Struct [`EthToolToUse`] which signals which function are ipmlemeted by this type.
+    const TO_USE: EthToolToUse;
+
+    /// Report driver/device information.  Should only set the
+    /// @driver, @version, @fw_version and @bus_info fields.  If not
+    /// implemented, the @driver and @bus_info fields will be filled in
+    /// according to the netdev's parent device.
+    #[allow(unused_variables)]
+    fn get_drvinfo(dev: &mut NetDevice<T>, info: &mut EthtoolDrvinfo) {}
+
+    /// Get the time stamping and PTP hardware clock capabilities.
+    /// Drivers supporting transmit time stamps in software should set this to
+    /// [`helpers::ethtool_op_get_ts_info`].
+    #[allow(unused_variables)]
+    fn get_ts_info(dev: &mut NetDevice<T>, info: &mut EthToolTsInfo) -> Result {
+        Err(Error::EINVAL)
+    }
+}
+
+/// Wrappes the [`bindings::ethtool_ts_info`] struct.
+#[repr(transparent)]
+pub struct EthToolTsInfo {
+    ptr: *const bindings::ethtool_ts_info,
+}
+
+impl SavedAsPointer for EthToolTsInfo {
+    type InternalType = bindings::ethtool_ts_info;
+
+    fn get_pointer(&self) -> *const Self::InternalType {
+        self.ptr
+    }
+
+    unsafe fn from_pointer(ptr: *const Self::InternalType) -> Self {
+        Self { ptr }
+    }
+}
+
+impl SavedAsPointerMut for EthToolTsInfo {
+    unsafe fn from_pointer_mut(ptr: *mut Self::InternalType) -> Self {
+        unsafe { Self::from_pointer(ptr as *const Self::InternalType) }
+    }
+}
+
+/// Wrappes the [`bindings::ethtool_drvinfo`] struct.
+pub struct EthtoolDrvinfo {
+    ptr: *const bindings::ethtool_drvinfo,
+}
+
+impl SavedAsPointer for EthtoolDrvinfo {
+    type InternalType = bindings::ethtool_drvinfo;
+
+    fn get_pointer(&self) -> *const Self::InternalType {
+        self.ptr
+    }
+
+    unsafe fn from_pointer(ptr: *const Self::InternalType) -> Self {
+        Self { ptr }
+    }
+}
+
+impl SavedAsPointerMut for EthtoolDrvinfo {
+    unsafe fn from_pointer_mut(ptr: *mut Self::InternalType) -> Self {
+        unsafe { Self::from_pointer(ptr as *const Self::InternalType) }
+    }
+}
+
+/// Helper functions for ethtool.
+pub mod helpers {
+    use super::*;
+
+    /// Get ts info for the device `dev`.
+    pub fn ethtool_op_get_ts_info<T: NetDeviceAdapter>(
+        dev: &mut NetDevice<T>,
+        info: &mut EthToolTsInfo,
+    ) -> Result {
+        // SAFETY: info.ptr is valid if info is valid
+        // SAFETY: Calling C function
```

Comment:
There should be only one `SAFETY` comment (please end with a period, too).

Also, what does it mean `if info is valid`? When can it be invalid? If a user can pass an invalid `info`, then is it UB? If so, why is `ethtool_op_get_ts_info` marked as safe?